### ISSUE TITLE
Preventing lgbm to stop too early.

### ISSUE THREAD

**Issue by lcrmorin** [microsoft/LightGBM#4791]
## Summary

I am currently working on a very noisy problem. I am working with a custom feval metric (the higher, the better) and use it for early stopping. 

However (when training without early stopping) the metric is often high by chance at the beginning of the training. Then it decrease to normal values as the training progress (say first 100 rounds), then increase again as the model learn over thousands of rounds. The metric often converge to values that are lower than the initial spike.

Adding early stopping cut the learning process n rounds after the initial spike, preventing the full learning process. I am trying to prevent early stopping to stop too early. I need some way to tell the early stopping to not start too early, maybe as a parameter in the early stopping callback.

## Motivation

This feature would be usefull for harder problem. Helping real-life adoption.

## Description

A parameter that would make the early stopping only start after n round, to prevent early over fitting. 

Ps: I am currently toying with feval too see if there isn't any tricky way to get this done. I am thinking about adding a penalty for earlier rounds, but can't find to access and use current round in feval function.

**Comment by StrikerRUS** [comment]
@lcrmorin Hey! Thank you for your interest in LightGBM!

> I am thinking about adding a penalty for earlier rounds, but can't find to access and use current round in feval function.

I think you can simply count iterations by yourself:

```python
import lightgbm as lgb

from sklearn.datasets import load_boston

X, y = load_boston(return_X_y=True)

curr_iter = 1

def do_not_stop_before_iter_10(preds, train_data):
    global curr_iter
    ret = curr_iter if curr_iter <= 10 else 10.  # just replace const 10. with your real metric value
    curr_iter += 1
    return 'do_not_stop_before_iter_10', ret, True

lgb_train = lgb.Dataset(X, y)
lgb_valid = lgb_train.create_valid(X, y)
gbm = lgb.train({'metric': 'custom'},
                lgb_train,
                num_boost_round=100,
                valid_sets=lgb_valid,
                feval=do_not_stop_before_iter_10,
                callbacks=[lgb.early_stopping(5)])

#> Early stopping, best iteration is:
#> [10]	valid_0's do_not_stop_before_iter_10: 10
```
